com.unity.multiplayer.mlapi/Runtime/Core/SnapshotSystem.cs

```diff
@@ -0,0 +1,412 @@
+        /// Registers the snapshot system for early updates
+        public SnapshotSystem()
+        {
+            this.RegisterNetworkUpdate(NetworkUpdateStage.EarlyUpdate);
+        }
+
+        /// <summary>
+        /// Dispose
+        /// </summary>
+        /// Unregisters the snapshot system from early updates
+        public void Dispose()
+        {
+            this.UnregisterNetworkUpdate(NetworkUpdateStage.EarlyUpdate);
+        }
+
+        public void NetworkUpdate(NetworkUpdateStage updateStage)
+        {
+            if (updateStage == NetworkUpdateStage.EarlyUpdate)
+            {
+                if (m_NetworkManager.IsServer)
+                {
+                    for (int i = 0; i < m_NetworkManager.ConnectedClientsList.Count; i++)
+                    {
+                        var clientId = m_NetworkManager.ConnectedClientsList[i].ClientId;
+                        SendSnapshot(clientId);
+                    }
+                }
+                else
+                {
+                    SendSnapshot(m_NetworkManager.ServerClientId);
+                }
+
+                DebugDisplayStore(m_Snapshot, "Entries");
+                DebugDisplayStore(m_ReceivedSnapshot, "Received Entries");
+            }
+        }
+
+        // todo --M1--
+        // for now, the full snapshot is always sent
+        // this will change significantly
+        /// <summary>
+        /// Send the snapshot to a specific client
+        /// </summary>
+        /// <param name="clientId">The client index to send to</param>
+        private void SendSnapshot(ulong clientId)
+        {
+            // Send the entry index and the buffer where the variables are serialized
+            using (var buffer = PooledNetworkBuffer.Get())
+            {
+                WriteIndex(buffer);
+                WriteBuffer(buffer);
+
+                m_NetworkManager.MessageSender.Send(clientId, NetworkConstants.SNAPSHOT_DATA,
+                    NetworkChannel.SnapshotExchange, buffer);
+                buffer.Dispose();
+            }
+        }
+
+        /// <summary>
+        /// Write the snapshot index to a buffer
+        /// </summary>
+        /// <param name="buffer">The buffer to write the index to</param>
+        private void WriteIndex(NetworkBuffer buffer)
+        {
+            using (var writer = PooledNetworkWriter.Get(buffer))
+            {
+                writer.WriteInt16((short)m_Snapshot.LastEntry);
+                for (var i = 0; i < m_Snapshot.LastEntry; i++)
+                {
+                    WriteEntry(writer, in m_Snapshot.Entries[i]);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Read the snapshot index from a buffer
+        /// Stores the entry. Allocates memory if needed. The actual buffer will be read later
+        /// </summary>
+        /// <param name="reader">The reader to read the index from</param>
+        private void ReadIndex(NetworkReader reader)
```

Comment:
The name read here could be misleading. If I'm using a "read" or a "get" method, I don't expect it to actually change data internally.

Reply:
From what I understand, this is called when you're reading the incoming data and adding it in the snapshot system right? How about "consume" or "process"?

Reply:
I kind-of disagree and would stick with `read` and `write`. Those are `private` methods. The very vast majority of the time spent looking at this code will be by developer working in the context of the snapshot system. They'll be knee-deep in sending/receiving data, managing states, etc. In this context `read` means gets data in from the network (change memory) and `write` means get data out to the network.

I also want to make it super clear that if you change the `read` part, you need to change the `write` part and vice-versa . I don't find it would be as obvious with consume/process. How strongly do you feel about this ?

Reply:
I don't, resolving